### ISSUE TITLE
expect() - provide option to set the test results to status different that OK and KO

### ISSUE THREAD

**Issue by fmancardi** [DevExpress/testcafe#6230]
Sometime while designing a test script, you are not able to fully complete the design/implementation.  
You can instruct your CI/CD system to do not run test scripts that have some METADATA, and this is OK.
But IMHO il will be useful to run the script anyway and if the t.expect() fails set the test status to BLOCKED or
any other custom string.

### What are you suggesting?
something like
t.expect(condition).ok({testStatusOnFailure: "BLOCKED"})

**Comment by wentwrong** [comment]
Could you please describe your use case in greater detail? From my perspective, you can achieve the same thing using [an assertion message](https://testcafe.io/documentation/402716/reference/test-api/testcontroller/expect/ok):
```js
await t.expect(condition).ok('Temporarily blocked by *reason*');
```
Or even skip this test until the tested functionality is implemented.

**Comment by fmancardi** [comment]
Hi
thanks for your answer.  
I'm using an script that analize the log generated by the pipeline that run the tests.
This script uses

✖ PPT-5942-Widget-Exists-Spedizioni-in-errore
 ✓ PPT-1147-EX15-contratto-modifica-tariffe-km-v1

 ✓ -> to write a succesful execution on [Testlink ](https://github.com/TestLinkOpenSourceTRMS/testlink-code) using the REST API

✖ -> to write failed

I would like to generate statistics also due to blocked test scripts, adding may be 
**BLOCKED** PPT-1147-EX15-contratto-modifica-tariffe-km-v1

That's my idea / request.

Hope now the request is a little bit clear

best regards

**Comment by VasilyStrelyaev** [comment]
Thank you for describing your scenario.
 
We designed our [custom reporters](https://testcafe.io/documentation/402810/guides/extend-testcafe/reporter-plugin) for sending test report data to external destinations. You can [access test metadata](https://testcafe.io/documentation/402790/reference/plugin-api/reporter#reporttestdone) from such a reporter and send additional info to analytics when a test with a certain metadata value fails.
 
So, parsing text reporters' output does not seem to be efficient.

**Comment by fmancardi** [comment]
I disagree with your approach, because does not solve the issue because you still have just two results OK o KO.
Parsing the reporter output is absolutely simpler that develop a new custom reporter.
There are several things IMHO that need to be improved regarding t.expect() or a similar mechanism.
Just think that in order to check if a selector exists and is visible the only way to use the automatic wait mechanism you need to use the expect() but you do not can choose to do not break test case execution if the expect fails.

If you need to implement logic inside your test scripts , a thing that is not a rare situation you need a little bit more freedom.

Please consider the request to add to the expect a parameter to set test case result other than PASSED or FAILED

best regards

**Comment by viktoria2506** [comment]
Assertions are designed so that a test fails if a condition fails. If you do not want the test to fail, handle the desired property from the selector in a different way.

With Selector.exists, auto-waiting doesn't work. We plan to work on this in the context of another issue - [exists ignores Selector timeout setting](https://github.com/DevExpress/testcafe/issues/5368). For now, you can use await Selector ().visible.

Adding a new test status is a technically challenging task. We believe that tests can be marked using a universal metadata engine. In most cases a test can be marked before it is launched – most users are satisfied with this way out. We also want to make it possible to dynamically change metadata - [Sending custom data to reporter ](https://github.com/DevExpress/testcafe/issues/3584)

**Comment by fmancardi** [comment]
>>  In most cases a test can be marked before it is launched – most users are satisfied with this way out.
it absolutely depends on how the people work.  
I use the metadata but for a different scope, metadata IMHO is intended to be used to say GO / NO Go with test execution, then this will exclude the test for any report.

All we know that when using manual test case execution, you do not have only two possible outcomes: Passed or Failed , but normally you can use other exec status like BLOCKED.

I can not say how challenging or difficult or costing can be adding a custom status on expect, I'm just repeating that from my point of view this can be an interesting feature to add.

I know also about the issues with exists method, and that you are working on it, and you can agree that fixing this is CRITICAL  because visible is not enough. But ... if I'm not wrong I can use the workaround explained in [exists ignores Selector timeout setting](https://github.com/DevExpress/testcafe/issues/5368), i.e. -> Selector timeouts allow you to retry obtaining an element's state (element properties) if the element doesn't exist. Then after the timeout if exists fails this means the element really does not exist.

Sending custom data to reporter can be useful but is open since July 2019.

**Comment by fmancardi** [comment]
Hi
I've done some simple experiments to have the BLOCKED status in the report output.  
My choice , following some advice in this discussion is using special string in the expect message.

    await t.expect(true).eql(false,'++BLOCKED++ This test is blocked');

Installed the https://www.npmjs.com/package/testcafe-reporter-mini, then changed the index.js file.  
Then I've got
 
 * PPT-7460-EX3-Tasto-Modifica-presenza-tab-Dati-aggiuntivi (screenshots:
 C:\development\testautomation\end-to-end\screenshots\2021-05-20_20-17-17\test-1\Chrome_90.0.4430.212_Windows_10\errors\1.png)
   AssertionError: ++BLOCKED++: expected true to deeply equal false + expected - actual -true
   +false

* -> symbol for blocked

another implementation can be display the string passed between ++ and ++.

 BLOCKED PPT-7460-EX3-Tasto-Modifica-presenza-tab-Dati-aggiuntivi (screenshots:

**Comment by AndreyBelym** [comment]
The POC implementation relies on a magic string in the error data. Such implementation cannot be considered secure enough, since a malicious user can trick the system by providing the magic string in an unexpected place (e.g. throw `new Error('++BLOCKED++')`).

Writing a proper implementation will be a lot harder. Also, we will need to write a quite big amount of automated tests for such change. Documenting this is also a serious task.

Moreover, we already have a lot of our own and third-party reporters. Make them support the new test status is a huge task. Also please keep in mind, that many report formats don't support the blocked status ([XUnit](https://xunit.net/docs/format-xml-v2#test), [JUnit](https://www.ibm.com/docs/en/z-open-unit-test/2.0.x?topic=SSZHNR_2.0.0/com.ibm.rsar.analysis.codereview.cobol.doc/topics/cac_useresults_junit.html))

TestCafe and its reporters are open source products, so it's completely fine if you fork a reporter and alter its output to be more suitable for your testing approach.

**Comment by fmancardi** [comment]
_The POC implementation relies on a magic string in the error data._  

there is no other way to provide the custom status.  It's no different to the suggestion to use a message instead of a custom status.

 _...cannot be considered secure enough,  _

this is just a poor's man workaround, the idea was to show that following your advice to use a custom reporter some
workaround was possible. Nothing more nothing less.

 _Writing a proper implementation will be a lot harder_  
It's clear for me, the POC was not done to say ' _... hey you said me is complex and I want to show you that is easy ..._'.
I've followed the advice to use in some way the available features -> i.e. the message

_Moreover, we already have a lot of our own and third-party reporters. Make them support the new test status is a huge task_
That's was not the idea, but to support this at least for the reporters you want to support, let's say testcafe-reporter-spec

Thanks for your time and support
Best regards
